### PR TITLE
feat(stkBNB certora): Certora rules for stakedBNB contract

### PR DESCRIPTION
**Pull request by vishal130195**
Add certora rules for stakedBNB contract.

### CODE REVIEW

File: certora/specs/StkBNB.spec [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+ rule SendSumOfFromAndToBalancesStaySame(address to, uint256 amount, bytes data) {
+    env e;
+    mathint sum = stakedBNBContract.balanceOf(e,e.msg.sender) + stakedBNBContract.balanceOf(e,to);
+    require sum < max_uint256;
+
+    uint256 totalSupplyBefore = totalSupply(e);
+    send(e, to, amount, data); 
+    uint256 totalSupplyAfter = totalSupply(e);
+
+    assert stakedBNBContract.balanceOf(e,e.msg.sender) + balanceOf(e,to) == sum;
```
Comment (hristo-grigorov) [inline certora/specs/StkBNB.spec:250]: Did you mean here:
`stakedBNBContract.balanceOf(e,to)` as in line 243
instead of:
`balanceOf(e,to)`?

**Review by hristo-grigorov — COMMENTED** [review]
The new rules look good! Good job!

File: certora/specs/StkBNB.spec [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+rule TransferSumOfFromAndToBalancesStaySame(address to, uint256 amount) {
+    env e;
+    mathint sum = stakedBNBContract.balanceOf(e,e.msg.sender) + stakedBNBContract.balanceOf(e,to);
+    require sum < max_uint256;
+
+    uint256 totalSupplyBefore = totalSupply(e);
+    transfer(e, to, amount); 
+    uint256 totalSupplyAfter = totalSupply(e);
+    // make use of send method too here
+    assert stakedBNBContract.balanceOf(e,e.msg.sender) + balanceOf(e,to) == sum;
```
[the remark this answers is not in the document]
Comment (vishal130195) [reply certora/specs/StkBNB.spec]: making use of balanceOf(e,to) every places now.
Done

File: certora/specs/StkBNB.spec [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+    env e1;
+	require e1.msg.sender == privileged;
+    bytes userData;
+    bytes operatorData;
+	storage initialStorage = lastStorage;
+    uint256 totalSupplyBefore = totalSupply(e1);
+	mint(e1, recipient, amount, userData, operatorData); // no revert
+	uint256 totalSupplyAfter1 = totalSupply(e1);
+
+    require(totalSupplyAfter1 > totalSupplyBefore);
```
[the remark this answers is not in the document]
Comment (vishal130195) [reply certora/specs/StkBNB.spec]: Done

File: certora/specs/StkBNB.spec [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+
+	env e2;
+	require e2.msg.sender != privileged;
+
+	mint@withrevert(e2, recipient, amount, userData, operatorData) at initialStorage;
+	bool secondSucceeded = !lastReverted;
+    uint256 totalSupplyAfter2 = totalSupply(e2);
+
+	assert  !secondSucceeded || (totalSupplyBefore == totalSupplyAfter2);
```
[the remark this answers is not in the document]
Comment (vishal130195) [reply certora/specs/StkBNB.spec]: Done

File: certora/specs/StkBNB.spec (hunk 1 above)
Comment (vishal130195) [reply certora/specs/StkBNB.spec]: Done

**Review by hristo-grigorov — COMMENTED** [review]
I left some comments. We can talk about them over our weekly call tomorrow.
In general, well-written rules.

File: certora/specs/StkBNB.spec [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+    assert seperate_balance == together_balance;
+}
+
+rule OnlyOwnerCanDestruct(){
+    env e;
+    selfDestruct(e);
+    assert e.msg.sender == getTimelockedAdminAddress(e) && paused(e) == true;
+}
+
+rule TransferFromSumOfFromAndToBalancesStaySame(address from, address to, uint256 amount) {
```
[the remark this answers is not in the document]
Comment (vishal130195) [reply certora/specs/StkBNB.spec]: Done

File: certora/specs/StkBNB.spec [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+}
+
+rule OnlyBurnerCanBurn(method f)filtered {f-> f.selector == burn(uint256, bytes).selector || f.selector == operatorBurn(address,uint256, bytes, bytes).selector} {
+    env e;
+    calldataarg args;
+    f(e, args);
+    assert(e.msg.sender == getStakePoolAddress());
+}
+
+rule OwnerCanChangeOwnership(address user){
```
[the remark this answers is not in the document]
Comment (vishal130195) [reply certora/specs/StkBNB.spec:69]: Done.

File: certora/specs/StkBNB.spec [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+}
+
+rule OnlyMinterCanMint(method f)filtered {f-> f.selector == mint(address,uint256, bytes, bytes).selector} {
+    env e;
+    calldataarg args;
+    f(e, args);
+    assert(e.msg.sender == getStakePoolAddress());
+}
+
+rule OnlyBurnerCanBurn(method f)filtered {f-> f.selector == burn(uint256, bytes).selector || f.selector == operatorBurn(address,uint256, bytes, bytes).selector} {
```
[the remark this answers is not in the document]
Comment (vishal130195) [reply certora/specs/StkBNB.spec]: Done

File: certora/specs/StkBNB.spec [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+
+ghost ghostGetStakePool() returns address {
+    axiom ghostGetStakePool() == stakePoolContract;
+}
+
+ghost ghostGetTimelockedAdmin() returns address {
+    axiom ghostGetTimelockedAdmin() == timelockedAdminContract;
+}
+
+rule OnlyMinterCanMint(method f)filtered {f-> f.selector == mint(address,uint256, bytes, bytes).selector} {
```
[the remark this answers is not in the document]
Comment (vishal130195) [reply certora/specs/StkBNB.spec]: Done

File: certora/specs/StkBNB.spec [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+    require sum < max_uint256;
+    
+    operatorSend(e, from, to, amount, data, operatorData); 
+    uint256 spenderBalanceAfter = balanceOf(e, e.msg.sender);
+
+    assert balanceOf(e, from) + balanceOf(e, to) == sum;
+    assert spenderBalanceBefore == spenderBalanceAfter;
+}
+
+rule AuthorizeOperatorCorrectness(address tokenHolder, address to, bytes data, bytes operatorData, uint256 amount) {
```
[the remark this answers is not in the document]
Comment (vishal130195) [reply certora/specs/StkBNB.spec:342]: As discussed, retain same and added check in previous rule.

**Review by hristo-grigorov — APPROVED** [review]
All looks good, well done

**Review by abhimanyusinghgaur — APPROVED** [review]

**Pull request merged**